File: src/frontend/expression_type.c

#include "expression_type.h"

bool expression_type_matches(struct expression_type *a, struct expression_type *b) {
	return a->access_type == b->access_type && a->width == b->width;
}

void expression_type_copy(struct expression_type *target, struct expression_type *source) {
	target->access_type = source->access_type;
	target->width = source->width;
}

File: src/frontend/expression_type.rs

#[derive(Debug, PartialEq, Clone)]
pub struct AccessType {
    pub readable: bool,
    pub writable: bool,
}

impl AccessType {
    pub fn r() -> AccessType {
        AccessType {
            readable: true,
            writable: false,
        }
    }

    pub fn w() -> AccessType {
        AccessType {
            readable: false,
            writable: true,
        }
    }

    pub fn rw() -> AccessType {
        AccessType {
            readable: true,
            writable: true,
        }
    }
}

#[derive(Debug, Clone)]
pub struct ExpressionType {
    pub access_type: AccessType,
    pub width: u64,
}
